sdk/src/main/java/com/qonversion/android/sdk/internal/QProductCenterManager.kt

```diff
@@ -245,19 +247,26 @@ internal class QProductCenterManager internal constructor(
     ) {
         loadProducts(object : QonversionProductsCallback {
             override fun onSuccess(products: Map<String, QProduct>) {
-                val storeIds = products.map { it.value.skuDetail?.sku }.filterNotNull()
-                repository.eligibilityForProductIds(
-                    storeIds,
-                    installDate,
-                    object : QonversionEligibilityCallback {
-                        override fun onSuccess(eligibilities: Map<String, QEligibility>) {
-                            val resultForRequiredProductIds =
-                                eligibilities.filter { it.key in productIds }
-                            callback.onSuccess(resultForRequiredProductIds)
-                        }
+                val result: MutableMap<String, QEligibility> = mutableMapOf()
+                products.forEach {
+                    val hasTrialOrIntroOffer = it.value.storeDetails?.hasTrialOrIntroOffer ?: false
+                    val isInApp = it.value.storeDetails?.productType == QProductType.InApp
+                    if (hasTrialOrIntroOffer) {
+                        result[it.key] = QEligibility(status = QIntroEligibilityStatus.Eligible)
+                    }
 
-                        override fun onError(error: QonversionError) = callback.onError(error)
-                    })
+                    if (isInApp) {
+                        result[it.key] = QEligibility(status = QIntroEligibilityStatus.NonIntroProduct)
+                    }
+
+                    if (it.value.type == QProductType.Trial && !hasTrialOrIntroOffer) {
+                        result[it.key] = QEligibility(status = QIntroEligibilityStatus.Ineligible)
+                    }
+                }
+
+                val resultForRequiredProductIds =
+                    result.filter { it.key in productIds }
```

Comment:
I think it would be better to filter products before calculation. Also better to use `map` as we convert each product to eligibility response.

```
products.filter { it.key in productIds }.map {
    ...
    QEligibility(status)
}
```